components/superchain-providers/src/ws.rs

```diff
@@ -1,171 +1,200 @@
+    provider::{EthereumProvider, Provider, StreamResponse, UniswapV2Provider, UniswapV3Provider},
+    types::{ethereum::BlockHeader, query::QueryOptions, uniswap_v2, uniswap_v3},
 };
 
+const WS_PATH: &str = "websocket";
+
 type WsMsg = Result<Vec<u8>>;
 type OperationMsg = (Operation, mpsc::UnboundedSender<WsMsg>);
 
-/// A Superchain WebSocket client
-pub struct Client {
+pub struct WsProvider {
     backend_tx: mpsc::Sender<OperationMsg>,
 }
 
-impl Client {
-    /// Create a new [`Client`]
-    pub async fn new<S>(websocket: WebSocketStream<S>) -> Self
+impl WsProvider {
+    async fn request<T>(&self, operation: Operation) -> StreamResponse<T>
     where
-        S: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+        T: serde::de::DeserializeOwned + 'static,
     {
-        let (tx, rx) = mpsc::channel(1024);
-        tokio::spawn(BackGroundWorker::new(websocket, rx).run());
+        let raw_data_stream = self.raw_request(operation).await?.boxed();
 
-        Self { backend_tx: tx }
+        let stream = csv_async::AsyncDeserializer::from_reader(raw_data_stream.into_async_read())
+            .into_deserialize::<T>()
+            .map_err(Error::from)
+            .boxed();
+
+        Ok(stream)
     }
 
-    /// Get the uniswap v2 pair created events for the provided `pairs_filter` within the specified
-    /// block range.
-    ///
-    /// A `pairs_filter` of `[]` or `None` will yield all `PairCreated` events. If one or more pair
-    /// hashes are specified, only these events will be returned (if present).
-    ///
-    /// A `from_block` of `None` will yield from the earliest indexed block (usually 0).
-    /// A `to_block_inc` of `None` will lead to a head following stream.
-    pub async fn get_pairs_created(
+    async fn raw_request(
         &self,
-        pairs_filter: impl IntoIterator<Item = H160>,
-        opts: QueryOptions,
-    ) -> Result<impl Stream<Item = Result<PairCreated>> + Send> {
+        operation: Operation,
+    ) -> Result<impl Stream<Item = Result<Vec<u8>, std::io::Error>> + Send> {
+        let (tx, rx) = mpsc::unbounded_channel();
+        self.backend_tx
+            .send((operation, tx))
+            .await
+            .map_err(|_| Error::BackendShutDown)?;
+
+        let raw_data_stream = futures::stream::unfold(rx, |mut rx| async move {
+            let res = rx.recv().await?;
+
+            match res {
+                Ok(data) => Some((Ok(data), rx)),
+                Err(err) => Some((Err(std::io::Error::new(std::io::ErrorKind::Other, err)), rx)),
+            }
+        });
+
+        Ok(raw_data_stream)
+    }
+}
+
+#[async_trait]
+impl Provider for WsProvider {
+    async fn new(
+        endpoint: String,
+        is_secure: bool,
+        username: Option<String>,
+        password: Option<String>,
+    ) -> Self {
+        let mut req = format!(
+            "{}://{}/{}",
+            if is_secure { "wss" } else { "ws" },
+            endpoint,
+            WS_PATH
```

Comment:
This will generate a warning with the latest `clippy`
```suggestion
            "{}://{endpoint}/{WS_PATH}",
            if is_secure { "wss" } else { "ws" },
```